packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts

```diff
@@ -0,0 +1,177 @@
+import { useMemo, useState, useEffect, useCallback } from 'react';
+import { useSearchParams, useRouter, usePathname } from 'next/navigation';
+import { useAtom } from 'jotai/react';
+import { UNDO } from 'jotai-history';
+import type {
+  FunnelJSON,
+  FunnelPosition,
+  NonChapterStep,
+} from '../types/funnel';
+import { FunnelStepTransitionType } from '../types/funnel';
+import type { TrackOnNavigate } from './useFunnelTracking';
+import {
+  funnelPositionAtom,
+  getFunnelStepByPosition,
+  funnelPositionHistoryAtom,
+} from '../store/funnelStore';
+
+interface UseFunnelNavigationProps {
+  funnel: FunnelJSON;
+  onNavigation: TrackOnNavigate;
+}
+
+type Chapters = Array<{ steps: number }>;
+type StepMap = Record<NonChapterStep['id'], { position: FunnelPosition }>;
+type NavigateFunction = (options: {
+  to: NonChapterStep['id'];
+  type?: FunnelStepTransitionType;
+}) => void;
+
+interface HeaderNavigation {
+  hasTarget: boolean;
+  navigate: () => void;
+}
+
+export interface UseFunnelNavigationReturn {
+  chapters: Chapters;
+  navigate: NavigateFunction;
+  position: FunnelPosition;
+  step: NonChapterStep;
+  back: HeaderNavigation;
+  skip: HeaderNavigation;
+}
+
+export const useFunnelNavigation = ({
+  funnel,
+  onNavigation,
+}: UseFunnelNavigationProps): UseFunnelNavigationReturn => {
+  const router = useRouter();
+  const searchParams = useSearchParams();
+  const pathname = usePathname();
+  const [stepTimerStart, setStepTimerStart] = useState<number>(0);
+  const [position, setPosition] = useAtom(funnelPositionAtom);
+  const [history, dispatchHistory] = useAtom(funnelPositionHistoryAtom);
+
+  const chapters: Chapters = useMemo(
+    () =>
+      funnel.steps
+        .filter((step) => 'steps' in step)
+        .map((chapter) => ({ steps: chapter.steps.length })),
+    [funnel],
+  );
+
+  const stepMap: StepMap = useMemo(() => {
+    const map: StepMap = {};
+    funnel.steps.forEach((chapter, chapterIndex) => {
+      if ('steps' in chapter) {
+        chapter.steps.forEach((step, stepIndex) => {
+          map[step.id] = {
+            position: {
+              chapter: chapterIndex,
+              step: stepIndex,
+            },
+          };
+        });
+      }
+    });
+    return map;
+  }, [funnel]);
+
+  const step: NonChapterStep = useMemo(
```

Comment:
What is non chapter step?

Reply:
Each step in funnel can be chapter or actual step. This is for every non chapter one

Reply:
> Rename it and move to `funnel.chapters.steps[stepIndex]`

Reply:
Updated ✔️